gprofiler/application_separators.py

```diff
@@ -0,0 +1,238 @@
+
+
+def get_cli_arg_by_name(args: List[str], arg_name: str, check_for_equals_arg: bool = False) -> str:
+    if arg_name in args:
+        return args[args.index(arg_name) + 1]
+
+    if check_for_equals_arg:
+        for arg in args:
+            arg_name, _, arg_val = arg.rpartition("=")
+            if arg_val is not None:
+                return arg_val
+
+    return _NON_AVAILABLE_ARG
+
+
+def get_cli_arg_by_index(args: List[str], index: int) -> str:
+    try:
+        return args[index]
+    except KeyError:
+        return _NON_AVAILABLE_ARG
+
+
+def append_python_module_to_proc_wd(process: Process, module_name: str) -> str:
+    return f'{process.cwd().replace("/", ".").strip(".")}.{module_name}'
+
+
+class GunicornApplicationSeparator(ApplicationSeparator):
+    @property
+    def app_prefix(self) -> str:
+        return "gunicorn"
+
+    def is_supported(self, process: Process) -> bool:
+        # Either gunicorn (for example: /usr/local/bin/gunicorn) or python that runs gunicorn
+        # (For example: /usr/local/bin/python /usr/local/bin/gunicorn)
+        return "gunicorn" in get_cli_arg_by_index(process.cmdline(), 0) or "gunicorn" in get_cli_arg_by_index(
+            process.cmdline(), 0
+        )
+
+    def get_application_name(self, process: Process) -> Optional[str]:
+        # As of gunicorn documentation the WSGI module name most probably will come from the cmdline and not from the
+        # config file / environment variables (they added the option to specify `wsgi_app` only in 20.1.0)
+        for arg in process.cmdline():
+            if ":" in arg:
+                if _IP_PORT_RE.match(arg):
+                    continue
+
+                return append_python_module_to_proc_wd(process, arg)
+
+        logger.warning(f"GunicornApplicationSeparator: matched against process {process} but couldn't find WSGI module")
+        return None
+
+
+class UwsgiApplicationSeparator(ApplicationSeparator):
+    @property
+    def app_prefix(self) -> str:
+        return "uwsgi"
+
+    def is_supported(self, process: Process) -> bool:
+        return "uwsgi" in process.cmdline()[0]
+
+    @staticmethod
+    def _find_wsgi_from_config_file(process: Process) -> Optional[str]:
+        config_file = None
+        for arg in process.cmdline():
+            if arg.endswith(".ini"):
+                config_file = arg
+
+        if config_file is None:
+            return None
+
+        if not os.path.isabs(config_file):
+            config_file = os.path.join(process.cwd(), config_file)
+
+        config = configparser.ConfigParser()
+        config.read(resolve_host_path(process, config_file))
+        try:
+            return config.get("uwsgi", "module")
+        except (configparser.NoSectionError, configparser.NoOptionError):
+            pass
+
+        return None
+
+    def get_application_name(self, process: Process) -> Optional[str]:
+        wsgi = get_cli_arg_by_name(process.cmdline(), "-w")
+        if wsgi is not None:
```

Comment:
In failure `get_cli_arg_by_name` returns an empty string. It applies to other usages of this functions as well.
IMO it's more common to return `None` in case of failure and therefor clearer and more intuitive than returning `""`. This opinion isn't very strict, but I do think it might save us some bugs in the future.

Reply:
And BTW `mypy` sucks

Reply:
It's not just empty string it's a sentinel of string type, I did it so all the checks for string contains will be very intuitive (without the need to check for returned type), returning Optional or Union type is shit both for mypy and both for the developer.

Reply:
It's a bad sentinel tho, since CPython interns strings...
```
In [9]: str() is str()                                                                                                                                                                                     
Out[9]: True

In [10]: str() is ''                                                                                                                                                                                       
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-10-bedcce6c87d1>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  str() is ''
Out[10]: True
```

An empty argument will be considered your sentinel. That won't do.

I suggest the following:
1. We can make `_get_cli_arg_by_name` and `_get_cli_arg_by_index` raise some `NoSuchArg` exception in the missing case. Upon being raised it will be caught in the high-level `get_application_name` and ignored - moving to the next app identifier. That's a fair solution IMO, but it works only if you use those 2 functions to test for "positive" cases and never for "negative" cases. which I'm not sure is the situation now.
2. Replace that with a real sentinel. You can see how I did it (with mypy compatibility) in `run_in_ns`: https://github.com/Granulate/granulate-utils/blob/1cfa3870036cecd3d2dfe9e9936376b0704846ba/granulate_utils/linux/ns.py#L71. It's fine IMO, not too much of a hassle.

I don't like exception-based programming all that much, and I prefer explicitness... So personally I'm up with option 2. But up to you.

Reply:
If returning a sentinel, why not just `None`?

Reply:
@Jongy I'm up with option 2, implemented it.
@YishaiZinkin `None` isn't nice because you should check type before comparing with other strings

Reply:
But if you replace with a "real" sentinel you'll need to check the type anyway

Reply:
nope, I declared a sentinel that derives from `str` and it works

Reply:
Hmm you could use `None` I suppose. But whatever